examples/multigpu_advtrain/model.py

```diff
@@ -0,0 +1,339 @@
+"""Lightweight model objects in TensorFlow.
+
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+import tensorflow as tf
+from utils import ordered_union
+
+from cleverhans.model import Model
+
+from tensorflow.python.training import moving_averages
+
+
+def clone_variable(name, x, trainable=False):
+    return tf.get_variable(name, shape=x.shape, dtype=x.dtype,
+                           trainable=trainable)
+
+
+class MLP(object):
+    """
+    An example of a bare bones multilayer perceptron (MLP) class.
+    """
+
+    def __init__(self, layers, input_shape, name=None, **kwargs):
+        self.layers = layers
+        self.input_shape = input_shape
+        self.name = name
+        if self.name is None:
+            self.name = 'MLP'
+        self.kwargs = kwargs
+
+    def get_params(self):
+        out = []
+        for layer in self.layers:
+            out = ordered_union(out, layer.get_params())
+        return out
+
+    def fprop(self, x, return_all=False, **kwargs):
+        kwargs.update(self.kwargs)
+        with tf.variable_scope(self.name):
+            states = []
+            for layer in self.layers:
+                x = layer.fprop(x, **kwargs)
+                assert x is not None
+                states.append(x)
+        if return_all:
+            return states
+        return x
+
+    def set_device(self, device_name):
+        for layer in self.layers:
+            layer.device_name = device_name
+
+    def __call__(self, x):
+        return self.fprop(x)
+
+    def create_sync_ops(self, host_device):
+        sync_ops = []
+        for layer in self.layers:
+            sync_ops += layer.create_sync_ops(host_device)
+        return sync_ops
+
+
+class MLP_probs(Model):
+    def __init__(self, mlp, name=None):
+        self.mlp = mlp
+
+    def get_layer_names(self):
+        return ['probs']
+
+    def set_device(self, device_name):
+        self.mlp.set_device(device_name)
+
+    def get_probs(self, x, **kwargs):
+        return self.mlp.fprop(x, **kwargs)
+
+    def fprop(self, x, **kwargs):
+        probs = self.mlp.fprop(x, **kwargs)
+        return {'probs': probs}
+
+    def create_sync_ops(self, host_device):
+        return self.mlp.create_sync_ops(host_device)
+
+
+class Layer(object):
```

Comment:
how hard would it be to use the Layer class from tutorial_models?

Reply:
Fixed in 714afc0. Now LayernGPU inherits from Layer in tutorial_models.